packages/async-flow/src/async-flow.js

```diff
@@ -0,0 +1,410 @@
+  const flowForOutcomeVowKey = outerZone.mapStore('flowForOutcomeVow', {
+    keyShape: M.remotable('vowishKey'),
+    valueShape: M.remotable('asyncFlow'),
+  });
+
+  /**
+   * @param {Zone} zone
+   * @param {string} tag
+   * @param {GuestAsyncFunc} guestAsyncFunc
+   * @param {{ startEager?: boolean }} [options]
+   */
+  const prepareAsyncFlowKit = (zone, tag, guestAsyncFunc, options = {}) => {
+    typeof guestAsyncFunc === 'function' ||
+      Fail`guestAsyncFunc must be a callable function ${guestAsyncFunc}`;
+    const {
+      // May change default to false, once instances reliably wake up
+      startEager = true,
+    } = options;
+
+    const internalMakeAsyncFlowKit = zone.exoClassKit(
+      tag,
+      AsyncFlowIKit,
+      (activationThis, activationArgs) => {
+        harden(activationThis);
+        harden(activationArgs);
+        const log = makeLogStore();
+        const bijection = makeWeakBijection();
+
+        return {
+          activationThis, // replay starts by reactivating with these
+          activationArgs, // replay starts by reactivating with these
+          log, // log to be accumulated or replayed
+          bijection, // membrane's guest-host mapping
+          outcomeKit: makeVowKit(), // outcome of activation as host vow
+          isDone: false, // persistently done
+        };
+      },
+      {
+        flow: {
+          restart(eager = startEager) {
+            const { state, facets } = this;
+            const {
+              activationThis,
+              activationArgs,
+              log,
+              bijection,
+              outcomeKit,
+            } = state;
+            const { flow, admin, wakeWatcher } = facets;
+
+            if (failures.has(flow)) {
+              failures.delete(flow);
+            }
+
+            !state.isDone ||
+              // separate line so I can set a breakpoint
+              Fail`Cannot restart a done flow ${flow}`;
+
+            admin.reset();
+            if (eager) {
+              eagerWakers.add(flow);
+            }
+
+            const wakeWatch = vowish => {
+              // Extra paranoid because we're getting
+              // "promise watcher must be a virtual object"
+              // in the general vicinity.
+              zone.isStorable(vowish) ||
+                Fail`vowish must be storable in this zone (usually, must be durable): ${vowish}`;
+              zone.isStorable(wakeWatcher) ||
+                Fail`wakeWatcher must be storable in this zone (usually, must be durable): ${wakeWatcher}`;
+              watch(vowish, wakeWatcher);
+            };
+            const panic = err => admin.panic(err);
+            const membrane = makeReplayMembrane(
+              log,
+              bijection,
+              vowTools,
+              wakeWatch,
+              panic,
+            );
+            const eph = tmp.for(flow);
+            eph.membrane = membrane;
+            const guestThis = membrane.hostToGuest(activationThis);
```

Comment:
From our conversation: let's make `this` inside the guest be a wrapper for the async flow itself. Passing the async flow instance to the guest allows it to reference the current invocation, which in the future would allow us to introduce features that lets an upgraded flow conditionally execute new behavior (e.g. ask question like "am I at a point where I'm still replaying") without violating the replay log (by calling a power outside of the membrane referencing the flow instance).

`asyncFlow` itself was never meant to support exo methods directly (the `this` context is not passable), and would always have required another wrapper on top, which can use the args just as much as the receiver to plumb the exo context through the membrane.

Reply:
Leaving this one open and undone for now. I suspect that we'll want to expose a new facet designed for this purpose, rather than exposing the existing `flow` facet.

Reply:
How do you feel about not making any such change in this PR, leaving this design issue for following PRs to address?

Reply:
At the very least I would like `this` to be undefined for now so no code starts relying on it.

Reply:
Done